File: library/src/main/java/com/jwoolston/android/uvc/interfaces/streaming/VideoFormat.java
package com.jwoolston.android.uvc.interfaces.streaming;

import android.support.annotation.NonNull;

import java.util.HashSet;
import java.util.Set;

/**
 * @author Jared Woolston (Jared.Woolston@gmail.com)
 */
public class VideoFormat<T extends VideoFrame> {

    protected int     formatIndex;
    protected int     numberFrames;
    protected int     defaultFrameIndex;
    protected int     aspectRatioX;
    protected int     aspectRatioY;
    protected byte    interlaceFlags;
    protected boolean copyProtect;

    private VideoColorMatchingDescriptor colorMatchingDescriptor;

    protected final Set<T> videoFrames = new HashSet<>();

    VideoFormat(@NonNull byte[] descriptor) throws IllegalArgumentException {

    }

    public void setColorMatchingDescriptor(VideoColorMatchingDescriptor descriptor) {
        colorMatchingDescriptor = descriptor;
    }

    public VideoColorMatchingDescriptor getColorMatchingDescriptor() {
        return colorMatchingDescriptor;
    }

    public int getFormatIndex() {
        return formatIndex;
    }

    public int getNumberFrames() {
        return numberFrames;
    }

    public int getDefaultFrameIndex() {
        return defaultFrameIndex;
    }

    @NonNull
    public Set<T> getVideoFrames() {
        return videoFrames;
    }

    public int getAspectRatioX() {
        return aspectRatioX;
    }

    public int getAspectRatioY() {
        return aspectRatioY;
    }

    public byte getInterlaceFlags() {
        return interlaceFlags;
    }

    public boolean isCopyProtect() {
        return copyProtect;
    }
}

File: library/src/main/java/com/jwoolston/android/uvc/interfaces/streaming/VideoFormat.kt
package com.jwoolston.android.uvc.interfaces.streaming

import com.jwoolston.android.uvc.streaming.VideoSampleFactory
import java.util.*

/**
 * @author Jared Woolston (Jared.Woolston@gmail.com)
 */
abstract class VideoFormat<T : VideoFrame> @Throws(IllegalArgumentException::class)
internal constructor(descriptor: ByteArray) {

    var formatIndex: Int = 0
        protected set
    var numberFrames: Int = 0
        protected set
    var defaultFrameIndex: Int = 0
        protected set
    var aspectRatioX: Int = 0
        protected set
    var aspectRatioY: Int = 0
        protected set
    var interlaceFlags: Byte = 0
        protected set
    var isCopyProtect: Boolean = false
        protected set

    var colorMatchingDescriptor: VideoColorMatchingDescriptor? = null

    val videoFrames: MutableSet<T> = HashSet()

    val defaultFrame: VideoFrame
        @Throws(IllegalStateException::class)
        get() {
            for (frame in videoFrames) {
                if (frame.frameIndex == defaultFrameIndex) {
                    return frame
                }
            }
            throw IllegalStateException("No default frame was found!")
        }

    abstract val displayName : String

    abstract fun getSampleFactory(maxPayload: Int, maxFrameSize: Int) : VideoSampleFactory
}
